rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -529,100 +741,429 @@ def slice_state_history(
+
+        # Collect all positive intent labels for a given state hash and add
+        # the set back to the singleton labels.
+        for positive_label_list in state_hash_to_label_list_instances.values():
+            # Get the set of positive labels associated with each state hash.
+            positive_label_set = set([labels[0] for labels in positive_label_list])
+
+            for labels in positive_label_list:
+                # Extend the singletone `labels` with the `positive_label_set`.
+                # Note that we need to filter out the redundant label `labels[0]`
+                # from `positive_label_set` so as not to add it twice.
+                filtered_label_set = filter(
+                    lambda label: label != labels[0], positive_label_set
+                )
+                labels.extend(filtered_label_set)
+
+        logger.debug(f"Created {len(example_states)} {self.LABEL_NAME} examples.")
+
+        return example_states, example_labels, example_entities
+
+    def _extract_examples(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        omit_unset_slots: bool = False,
+        ignore_action_unlikely_intent: bool = False,
+    ) -> Iterator[Tuple[List[State], List[Text], List[Dict[Text, Any]]]]:
+        """Creates an iterator over training examples from a tracker.
+
+        Args:
+            trackers: The tracker from which to extract training examples.
+            domain: The domain of the training data.
+            omit_unset_slots: If `True` do not include the initial values of slots.
+            ignore_action_unlikely_intent: Whether to remove `action_unlikely_intent`
+                from training states.
+
+        Returns:
+            An iterator over example states, labels, and entity data.
+        """
+        tracker_states = self._create_states(
+            tracker, domain, omit_unset_slots=omit_unset_slots
+        )
+        events = tracker.applied_events()
+
+        if ignore_action_unlikely_intent:
+            tracker_states = self._remove_action_unlikely_intent_from_states(
+                tracker_states
+            )
+            events = self._remove_action_unlikely_intent_from_events(events)
+
+        label_index = 0
+        for event in events:
+
+            if isinstance(event, ActionExecuted):
+                label_index += 1
+
+            elif isinstance(event, UserUttered):
+
+                sliced_states = self.slice_state_history(
+                    tracker_states[:label_index], self.max_history
+                )
+                label = [event.intent_name or event.text]
+                entities = [{}]
+
+                yield sliced_states, label, entities
+
+    @staticmethod
+    def _cleanup_last_user_state_with_action_listen(
+        trackers_as_states: List[List[State]],
+    ) -> List[List[State]]:
+        """Deletes the last user state where previous action is `action_listen`.
```

Comment:
@wochinge  I see your point about the docstring not being helpful but then should we couple the why of the function with the function itself. IMO info on why a function is called somewhere should be in the place where it is called. Moreover, this method is a static method (more like a utility method) and it can be used independently anywhere. It may not be clear that every call to the function is for the same reason.

Reply:
I think you would only want to call this function if you are predicting something about the state you are removing, so including that part of the why, how does this read to y'all:
```
Removes the last state in a tracker if the previous action is action_listen.

States with the previous action equal to action_listen correspond to states with
a new user intent. This information is what IntentTEDPolicy is trying to predict
so it needs to be removed before obtaining a prediction.
```

Reply:
Cool, sounds good to me. I've already renamed everything to `UnexpecTEDIntentPolicy` on this branch, so please change that in the description :)

Reply:
Addressed in https://github.com/RasaHQ/rasa/pull/9005. Will resolve this conversation when that PR has been merged into `intent-ted`.